plugins/terminal.koplugin/main.lua

```diff
@@ -0,0 +1,96 @@
+
+local DataStorage = require("datastorage")
+local Font = require("ui/font")
+local InfoMessage = require("ui/widget/infomessage")
+local InputDialog = require("ui/widget/inputdialog")
+local KeyValuePage = require("ui/widget/keyvaluepage")
+local Screen = require("device").screen
+local UIManager = require("ui/uimanager")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local logger = require("logger")
+local util = require("ffi/util")
+local _ = require("gettext")
+
+local Terminal = WidgetContainer:new{
+    name = "terminal",
+    dump_file = util.realpath(DataStorage:getDataDir()) .. "/terminal_output.txt",
+}
+
+function Terminal:init()
+    self.ui.menu:registerToMainMenu(self)
+end
+
+function Terminal:start()
+    self.input = InputDialog:new{
+        title =  _("Enter a command and press \"Execute\""),
+        text_height = Screen:getHeight() * 0.4,
+        input_type = "string",
+        buttons = {{{
+            text = _("Cancel"),
+            callback = function()
+                UIManager:close(self.input)
+            end,
+        }, {
+            text = _("Execute"),
+            is_enter_default = true,
+            callback = function()
+                UIManager:close(self.input)
+                self:execute()
+            end,
+        }}},
+    }
+    self.input:onShowKeyboard()
+    UIManager:show(self.input)
+end
+
+function Terminal:execute()
+    local command = self.input:getInputText()
+    UIManager:show(InfoMessage:new{
+        text = _("Executing ..."),
+        timeout = 0.1,
+    })
+    UIManager:forceRePaint()
+    local std_out = io.popen(command)
+    local entries = { command }
+    if std_out then
+        while true do
+            local line = std_out:read()
+            if line == nil then break end
+            table.insert(entries, line)
+        end
+        std_out:close()
+    else
+        table.insert(entries, _("Failed to execute command."))
+    end
+    self:dump(entries)
+    table.insert(entries, _("Output will also be dumped to."))
+    table.insert(entries, self.dump_file)
+    UIManager:show(KeyValuePage:new{
```

Comment:
If I have not misunderstood, scrolltext displays everything without preserving the lines, which is the most important reason I do not prefer to use it in terminal.
Comparing two output formats,
`
drwxrwxr-x  6 hzj_jie hzj_jie     4096 Mar 20 13:46 frontend/
drwxrwxr-x  9 hzj_jie hzj_jie     4096 Mar 20 14:11 .git/
drwxrwxr-x  2 hzj_jie hzj_jie     4096 Mar 17  2016 .github/
-rw-rw-r--  1 hzj_jie hzj_jie      778 Feb 25 17:08 .gitignore
-rw-rw-r--  1 hzj_jie hzj_jie      486 Jun 28  2016 .gitmodules
-rwxrwxr-x  1 hzj_jie hzj_jie    13527 Mar 20 13:46 kodev*
drwxr-xr-x  3 hzj_jie hzj_jie     4096 Sep 16  2016 koreader-android-arm-linu >>
-rw-rw-r--  1 hzj_jie hzj_jie 19826400 Sep 16  2016 koreader-android-arm-linu >>
drwxr-xr-x  3 hzj_jie hzj_jie     4096 Jan  6 01:59 koreader-emulator-x86_64-linu >>
drwxr-xr-x  3 hzj_jie hzj_jie     4096 Mar  1 22:52 koreader-kindle-arm-linu >>
-rw-rw-r--  1 hzj_jie hzj_jie 22124662 Dec 30 20:33 koreader-kindle-arm-linu >>
-rw-rw-r--  1 hzj_jie hzj_jie 22319864 Dec 30 20:33 koreader-kindle-arm-linu >>
`
and
`
drwxrwxr-x  6 hzj_jie hzj_jie     4096 Mar 20 13:46 frontend/
drwxrwxr-x  9 hzj_jie hzj_jie     4096 Mar 20 14:11 .git/
drwxrwxr-x  2 hzj_jie hzj_jie     4096 Mar 17  2016 .github/
-rw-rw-r--  1 hzj_jie hzj_jie      778 Feb 25 17:08 .gitignore
-rw-rw-r--  1 hzj_jie hzj_jie      486 Jun 28  2016 .gitmodules
-rwxrwxr-x  1 hzj_jie hzj_jie    13527 Mar 20 13:46 kodev*
drwxr-xr-x  3 hzj_jie hzj_jie     4096 Sep 16  2016 koreader-android-arm-linux-androideabi/
-rw-rw-r--  1 hzj_jie hzj_jie 19826400 Sep 16  2016 koreader-android-arm-linux-androideabi-v2015.11-653-g50d5d92.apk
drwxr-xr-x  3 hzj_jie hzj_jie     4096 Jan  6 01:59 koreader-emulator-x86_64-linux-gnu/
drwxr-xr-x  3 hzj_jie hzj_jie     4096 Mar  1 22:52 koreader-kindle-arm-linux-gnueabi/
-rw-rw-r--  1 hzj_jie hzj_jie 22124662 Dec 30 20:33 koreader-kindle-arm-linux-gnueabi-v2015.11-793-g7ef26d0.targz
-rw-rw-r--  1 hzj_jie hzj_jie 22319864 Dec 30 20:33 koreader-kindle-arm-linux-gn
`

I think the first one is more readable. Though both of them are not ideal.